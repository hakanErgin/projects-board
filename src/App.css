@import '~antd/dist/antd.css';

body {
  font-family: 'Open Sans';
  background-color: rgb(243, 245, 251);
}

.App {
  padding-top: 10%;
  display: flex;
  flex-direction: column;
  align-items: center;
}

.ProjectList {
  padding: 5px 25px 10px 25px;
  background-color: white;
  min-width: 475px;
  border-radius: 20px;
  box-shadow: 0px 4px 10px 0px rgb(211, 210, 210);
}

.ant-list-split .ant-list-header {
  border-bottom: 0px;
}

.HeaderTitle {
  margin-bottom: 0px;
  font-weight: 600;
}

.ant-list-item-meta-title {
  margin-bottom: 0px;
}

.Header {
  display: flex;
  justify-content: space-between;
  align-items: center;
  border-bottom: 0px;
  font-size: 17px;
  margin-bottom: 0px;
}

.Icon {
  font-size: 21px;
  color: rgb(53, 53, 53);
}

.ProjectTitles {
  font-size: 20px;
  margin-bottom: 0px;
}

.ProjectDescriptions {
  font-size: 14px;
  font-weight: 600;
  margin-bottom: 2px;
}

.ant-list-item {
  padding: 8px 0;
}

.WelcomeText {
  width: 75%;
  font-weight: 1000;
  font-size: 36px;
  line-height: 1.25;
}

.WelcomeTextContainer {
  width: 475px;
}
