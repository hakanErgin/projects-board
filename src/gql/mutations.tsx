import { gql } from '@apollo/client';

export const UPDATE_PROJECT = gql`
  mutation($id: ID!, $name: String, $enterprise_id: ID) {
    updateProject(id: $id, name: $name, enterprise_id: $enterprise_id) {
      id
      name
      enterprise_id
      Enterprise {
        id
        name
      }
      Users {
        id
      }
    }
  }
`;

export const CREATE_PROJECT = gql`
  mutation($id: ID!, $name: String!, $enterprise_id: ID!) {
    createProject(id: $id, name: $name, enterprise_id: $enterprise_id) {
      id
    }
  }
`;

export const REMOVE_PROJECT = gql`
  mutation($id: ID!) {
    removeProject(id: $id)
  }
`;

export const UPDATE_USER = gql`
  mutation($id: ID!, $project_id: ID) {
    updateUser(id: $id, project_id: $project_id) {
      id
      project_id
    }
  }
`;
