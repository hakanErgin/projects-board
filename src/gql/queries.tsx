import { gql } from '@apollo/client';

export const GET_ENTERPRISES = gql`
  query {
    allEnterprises {
      id
      name
    }
  }
`;

export const GET_PROJECT = gql`
  query($id: ID!) {
    Project(id: $id) {
      id
      name
      Enterprise {
        id
        name
      }
    }
  }
`;

export const GET_PROJECTS = gql`
  query {
    allProjects {
      id
      name
      Users {
        id
      }
    }
  }
`;

export const GET_ALL_USERS = gql`
  query {
    allUsers {
      id
      email
      first_name
      last_name
    }
  }
`;

export const GET_USERS_BY_PROJECT = gql`
  query($id: ID!) {
    Project(id: $id) {
      Users {
        id
        first_name
        last_name
        avatar
        email
        project_id
      }
    }
  }
`;
