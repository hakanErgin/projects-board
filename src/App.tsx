import React, { useState } from 'react';
import { useQuery, useMutation } from '@apollo/client';
import { List } from 'antd';
import {
  UserSwitchOutlined,
  EditOutlined,
  DeleteOutlined,
  PlusOutlined,
} from '@ant-design/icons';
import { AddProjectModal, EditProjectModal, UserModal } from './components';
import { GET_PROJECTS, REMOVE_PROJECT } from './gql';
import './App.css';
import { Project } from './types';

function App() {
  const [selectedProjectId, setselectedProjectId] = useState<string>('');

  // modal visibilty states
  const [isAddProjectModalVisible, setIsAddProjectModalVisible] = useState(
    false
  );
  const [isUserModalVisible, setIsUserModalVisible] = useState(false);
  const [isEditProjectModalVisible, setIsEditProjectModalVisible] = useState(
    false
  );

  // gql api hooks
  const {
    loading: projectsLoading,
    error: projectsError,
    data: projectsData,
  } = useQuery(GET_PROJECTS);
  const [removeProject] = useMutation(REMOVE_PROJECT, {
    errorPolicy: 'ignore',
  });

  // making sure fetched data is ready
  if (projectsLoading) return <p id="loading">Loading...</p>;
  if (projectsError) return <p>Error</p>;

  // modal visibility functions
  function showAddProjectModal() {
    setIsAddProjectModalVisible(true);
  }
  function showUserModal(projectId: string) {
    setselectedProjectId(projectId);
    setIsUserModalVisible(true);
  }
  function showEditProjectModal(projectId: string) {
    setselectedProjectId(projectId);
    setIsEditProjectModalVisible(true);
  }

  // component logic functions
  function handleRemoveProject(projectId: string) {
    removeProject({
      variables: {
        id: projectId,
      },
      // updating cache instead of having to refetch
      update: (cache) => {
        const projects: any = cache.readQuery({
          query: GET_PROJECTS,
        });
        const newProjects = projects.allProjects.filter(
          (project: Project) => project.id !== projectId
        );
        cache.writeQuery({
          query: GET_PROJECTS,
          data: { allProjects: { newProjects } },
        });
      },
    });
  }

  return (
    <div className="App">
      {isAddProjectModalVisible && (
        <AddProjectModal
          isAddProjectModalVisible={isAddProjectModalVisible}
          setIsAddProjectModalVisible={setIsAddProjectModalVisible}
        />
      )}
      {isUserModalVisible && (
        <UserModal
          selectedProjectId={selectedProjectId}
          isUserModalVisible={isUserModalVisible}
          setIsUserModalVisible={setIsUserModalVisible}
        />
      )}
      {isEditProjectModalVisible && (
        <EditProjectModal
          selectedProjectId={selectedProjectId}
          isEditProjectModalVisible={isEditProjectModalVisible}
          setIsEditProjectModalVisible={setIsEditProjectModalVisible}
        />
      )}
      <div className={'WelcomeTextContainer'}>
        <h1 className={'WelcomeText'}>Welcome to your projects board</h1>
      </div>
      <List
        header={[
          <div key={'header'} className={'Header'}>
            <h2 className={'HeaderTitle'}>Projects</h2>
            <PlusOutlined className={'Icon'} onClick={showAddProjectModal} />
          </div>,
        ]}
        className="ProjectList"
        itemLayout="horizontal"
        dataSource={projectsData.allProjects}
        renderItem={(item: Project) => {
          return (
            <List.Item
              className={'ListItem'}
              actions={[
                <UserSwitchOutlined
                  className={'Icon'}
                  onClick={() => showUserModal(item.id)}
                />,
                <EditOutlined
                  className={'Icon'}
                  onClick={() => showEditProjectModal(item.id)}
                />,
                <DeleteOutlined
                  className={'Icon'}
                  onClick={() => handleRemoveProject(item.id)}
                />,
              ]}
            >
              <List.Item.Meta
                className={'ListItem'}
                title={<p className={'ProjectTitles'}>{item.name}</p>}
                description={
                  <p className={'ProjectDescriptions'}>
                    {item.Users.length + ' collaborators'}
                  </p>
                }
              />
            </List.Item>
          );
        }}
      />
    </div>
  );
}

export default App;
