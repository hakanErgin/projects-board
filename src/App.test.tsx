import React from 'react';
import renderer, { act } from 'react-test-renderer';
import App from './App';
import { MockedProvider } from '@apollo/client/testing';
import { GET_PROJECTS } from './gql/';
import wait from 'waait';

const projectsMock = {
  request: {
    query: GET_PROJECTS,
  },
  result: {
    data: {
      allProjects: [
        {
          id: '1',
          name: 'Sub-Ex',
          Users: [
            {
              id: '1',
            },
          ],
        },
        {
          id: '2',
          name: 'Second',
          Users: [
            {
              id: '2',
            },
          ],
        },
      ],
    },
  },
};

it('renders loading state', () => {
  const component = renderer.create(
    <MockedProvider mocks={[projectsMock]} addTypename={false}>
      <App />
    </MockedProvider>
  );
  const tree: any = component.toJSON();
  expect(tree.children).toContain('Loading...');
});
it('renders correctly with mocked data', async () => {
  const component = renderer.create(
    <MockedProvider mocks={[projectsMock]} addTypename={false}>
      <App />
    </MockedProvider>
  );
  await act(wait);
  expect(component).toMatchSnapshot();
});
it('should show 2 elems fetched', async () => {
  const component = renderer.create(
    <MockedProvider mocks={[projectsMock]} addTypename={false}>
      <App />
    </MockedProvider>
  );
  await act(wait);
  const ul = component.root.findByType('ul');
  expect(ul.children.length).toBe(2);
});
