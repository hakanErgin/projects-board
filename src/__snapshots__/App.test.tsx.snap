// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly with mocked data 1`] = `
<div
  className="App"
>
  <div
    className="WelcomeTextContainer"
  >
    <h1
      className="WelcomeText"
    >
      Welcome to your projects board
    </h1>
  </div>
  <div
    className="ant-list ant-list-split ProjectList"
  >
    <div
      className="ant-list-header"
    >
      <div
        className="Header"
      >
        <h2
          className="HeaderTitle"
        >
          Projects
        </h2>
        <span
          aria-label="plus"
          className="anticon anticon-plus Icon"
          onClick={[Function]}
          role="img"
          tabIndex={-1}
        >
          <svg
            aria-hidden="true"
            className=""
            data-icon="plus"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="64 64 896 896"
            width="1em"
          >
            <defs>
              <style />
            </defs>
            <path
              d="M482 152h60q8 0 8 8v704q0 8-8 8h-60q-8 0-8-8V160q0-8 8-8z"
            />
            <path
              d="M176 474h672q8 0 8 8v60q0 8-8 8H176q-8 0-8-8v-60q0-8 8-8z"
            />
          </svg>
        </span>
      </div>
    </div>
    <div
      className="ant-spin-nested-loading"
    >
      <div
        className="ant-spin-container"
      >
        <ul
          className="ant-list-items"
        >
          <li
            className="ant-list-item ListItem"
          >
            <div
              className="ant-list-item-meta ListItem"
            >
              <div
                className="ant-list-item-meta-content"
              >
                <h4
                  className="ant-list-item-meta-title"
                >
                  <p
                    className="ProjectTitles"
                  >
                    Sub-Ex
                  </p>
                </h4>
                <div
                  className="ant-list-item-meta-description"
                >
                  <p
                    className="ProjectDescriptions"
                  >
                    1 collaborators
                  </p>
                </div>
              </div>
            </div>
            <ul
              className="ant-list-item-action"
            >
              <li>
                <span
                  aria-label="user-switch"
                  className="anticon anticon-user-switch Icon"
                  onClick={[Function]}
                  role="img"
                  tabIndex={-1}
                >
                  <svg
                    aria-hidden="true"
                    className=""
                    data-icon="user-switch"
                    fill="currentColor"
                    focusable="false"
                    height="1em"
                    viewBox="64 64 896 896"
                    width="1em"
                  >
                    <defs>
                      <style />
                    </defs>
                    <path
                      d="M759 335c0-137-111-248-248-248S263 198 263 335c0 82.8 40.6 156.2 103 201.2-.4.2-.7.3-.9.4-44.7 18.9-84.8 46-119.3 80.6a373.42 373.42 0 00-80.4 119.5A373.6 373.6 0 00136 874.8a8 8 0 008 8.2h59.9c4.3 0 7.9-3.5 8-7.8 2-77.2 32.9-149.5 87.6-204.3C356 614.2 431 583 511 583c137 0 248-111 248-248zM511 507c-95 0-172-77-172-172s77-172 172-172 172 77 172 172-77 172-172 172zm105 221h264c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H703.5l47.2-60.1a8.1 8.1 0 001.7-4.9c0-4.4-3.6-8-8-8h-72.6c-4.9 0-9.5 2.3-12.6 6.1l-68.5 87.1c-4.4 5.6-6.8 12.6-6.8 19.8.1 17.7 14.4 32 32.1 32zm240 64H592c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h176.5l-47.2 60.1a8.1 8.1 0 00-1.7 4.9c0 4.4 3.6 8 8 8h72.6c4.9 0 9.5-2.3 12.6-6.1l68.5-87.1c4.4-5.6 6.8-12.6 6.8-19.8-.1-17.7-14.4-32-32.1-32z"
                    />
                  </svg>
                </span>
                <em
                  className="ant-list-item-action-split"
                />
              </li>
              <li>
                <span
                  aria-label="edit"
                  className="anticon anticon-edit Icon"
                  onClick={[Function]}
                  role="img"
                  tabIndex={-1}
                >
                  <svg
                    aria-hidden="true"
                    className=""
                    data-icon="edit"
                    fill="currentColor"
                    focusable="false"
                    height="1em"
                    viewBox="64 64 896 896"
                    width="1em"
                  >
                    <path
                      d="M257.7 752c2 0 4-.2 6-.5L431.9 722c2-.4 3.9-1.3 5.3-2.8l423.9-423.9a9.96 9.96 0 000-14.1L694.9 114.9c-1.9-1.9-4.4-2.9-7.1-2.9s-5.2 1-7.1 2.9L256.8 538.8c-1.5 1.5-2.4 3.3-2.8 5.3l-29.5 168.2a33.5 33.5 0 009.4 29.8c6.6 6.4 14.9 9.9 23.8 9.9zm67.4-174.4L687.8 215l73.3 73.3-362.7 362.6-88.9 15.7 15.6-89zM880 836H144c-17.7 0-32 14.3-32 32v36c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-36c0-17.7-14.3-32-32-32z"
                    />
                  </svg>
                </span>
                <em
                  className="ant-list-item-action-split"
                />
              </li>
              <li>
                <span
                  aria-label="delete"
                  className="anticon anticon-delete Icon"
                  id="deleteButton"
                  onClick={[Function]}
                  role="img"
                  tabIndex={-1}
                >
                  <svg
                    aria-hidden="true"
                    className=""
                    data-icon="delete"
                    fill="currentColor"
                    focusable="false"
                    height="1em"
                    viewBox="64 64 896 896"
                    width="1em"
                  >
                    <path
                      d="M360 184h-8c4.4 0 8-3.6 8-8v8h304v-8c0 4.4 3.6 8 8 8h-8v72h72v-80c0-35.3-28.7-64-64-64H352c-35.3 0-64 28.7-64 64v80h72v-72zm504 72H160c-17.7 0-32 14.3-32 32v32c0 4.4 3.6 8 8 8h60.4l24.7 523c1.6 34.1 29.8 61 63.9 61h454c34.2 0 62.3-26.8 63.9-61l24.7-523H888c4.4 0 8-3.6 8-8v-32c0-17.7-14.3-32-32-32zM731.3 840H292.7l-24.2-512h487l-24.2 512z"
                    />
                  </svg>
                </span>
              </li>
            </ul>
          </li>
          <li
            className="ant-list-item ListItem"
          >
            <div
              className="ant-list-item-meta ListItem"
            >
              <div
                className="ant-list-item-meta-content"
              >
                <h4
                  className="ant-list-item-meta-title"
                >
                  <p
                    className="ProjectTitles"
                  >
                    Second
                  </p>
                </h4>
                <div
                  className="ant-list-item-meta-description"
                >
                  <p
                    className="ProjectDescriptions"
                  >
                    1 collaborators
                  </p>
                </div>
              </div>
            </div>
            <ul
              className="ant-list-item-action"
            >
              <li>
                <span
                  aria-label="user-switch"
                  className="anticon anticon-user-switch Icon"
                  onClick={[Function]}
                  role="img"
                  tabIndex={-1}
                >
                  <svg
                    aria-hidden="true"
                    className=""
                    data-icon="user-switch"
                    fill="currentColor"
                    focusable="false"
                    height="1em"
                    viewBox="64 64 896 896"
                    width="1em"
                  >
                    <defs>
                      <style />
                    </defs>
                    <path
                      d="M759 335c0-137-111-248-248-248S263 198 263 335c0 82.8 40.6 156.2 103 201.2-.4.2-.7.3-.9.4-44.7 18.9-84.8 46-119.3 80.6a373.42 373.42 0 00-80.4 119.5A373.6 373.6 0 00136 874.8a8 8 0 008 8.2h59.9c4.3 0 7.9-3.5 8-7.8 2-77.2 32.9-149.5 87.6-204.3C356 614.2 431 583 511 583c137 0 248-111 248-248zM511 507c-95 0-172-77-172-172s77-172 172-172 172 77 172 172-77 172-172 172zm105 221h264c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H703.5l47.2-60.1a8.1 8.1 0 001.7-4.9c0-4.4-3.6-8-8-8h-72.6c-4.9 0-9.5 2.3-12.6 6.1l-68.5 87.1c-4.4 5.6-6.8 12.6-6.8 19.8.1 17.7 14.4 32 32.1 32zm240 64H592c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h176.5l-47.2 60.1a8.1 8.1 0 00-1.7 4.9c0 4.4 3.6 8 8 8h72.6c4.9 0 9.5-2.3 12.6-6.1l68.5-87.1c4.4-5.6 6.8-12.6 6.8-19.8-.1-17.7-14.4-32-32.1-32z"
                    />
                  </svg>
                </span>
                <em
                  className="ant-list-item-action-split"
                />
              </li>
              <li>
                <span
                  aria-label="edit"
                  className="anticon anticon-edit Icon"
                  onClick={[Function]}
                  role="img"
                  tabIndex={-1}
                >
                  <svg
                    aria-hidden="true"
                    className=""
                    data-icon="edit"
                    fill="currentColor"
                    focusable="false"
                    height="1em"
                    viewBox="64 64 896 896"
                    width="1em"
                  >
                    <path
                      d="M257.7 752c2 0 4-.2 6-.5L431.9 722c2-.4 3.9-1.3 5.3-2.8l423.9-423.9a9.96 9.96 0 000-14.1L694.9 114.9c-1.9-1.9-4.4-2.9-7.1-2.9s-5.2 1-7.1 2.9L256.8 538.8c-1.5 1.5-2.4 3.3-2.8 5.3l-29.5 168.2a33.5 33.5 0 009.4 29.8c6.6 6.4 14.9 9.9 23.8 9.9zm67.4-174.4L687.8 215l73.3 73.3-362.7 362.6-88.9 15.7 15.6-89zM880 836H144c-17.7 0-32 14.3-32 32v36c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-36c0-17.7-14.3-32-32-32z"
                    />
                  </svg>
                </span>
                <em
                  className="ant-list-item-action-split"
                />
              </li>
              <li>
                <span
                  aria-label="delete"
                  className="anticon anticon-delete Icon"
                  id="deleteButton"
                  onClick={[Function]}
                  role="img"
                  tabIndex={-1}
                >
                  <svg
                    aria-hidden="true"
                    className=""
                    data-icon="delete"
                    fill="currentColor"
                    focusable="false"
                    height="1em"
                    viewBox="64 64 896 896"
                    width="1em"
                  >
                    <path
                      d="M360 184h-8c4.4 0 8-3.6 8-8v8h304v-8c0 4.4 3.6 8 8 8h-8v72h72v-80c0-35.3-28.7-64-64-64H352c-35.3 0-64 28.7-64 64v80h72v-72zm504 72H160c-17.7 0-32 14.3-32 32v32c0 4.4 3.6 8 8 8h60.4l24.7 523c1.6 34.1 29.8 61 63.9 61h454c34.2 0 62.3-26.8 63.9-61l24.7-523H888c4.4 0 8-3.6 8-8v-32c0-17.7-14.3-32-32-32zM731.3 840H292.7l-24.2-512h487l-24.2 512z"
                    />
                  </svg>
                </span>
              </li>
            </ul>
          </li>
        </ul>
      </div>
    </div>
  </div>
</div>
`;
